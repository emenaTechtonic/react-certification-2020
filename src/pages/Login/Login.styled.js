import styled from 'styled-components'

const StyledLogin = styled.div.attrs({
  role: 'login',
})`
  background-color: white;
  border-radius: 1em;
  .login-wrapper {
    display: flex;
    flex-direction: column;
    align-items: center;
  }
  .login {
    width: 300px;
  }

  .login h1 {
    text-align: center;
    letter-spacing: -1px;
  }

  .login-form {
    display: flex;
    flex-direction: column;
    align-items: center;
  }

  .form-group {
    width: 100%;
    display: flex;
    flex-direction: column;
    margin-bottom: 1rem;
  }

  .form-group strong {
    display: block;
    font-weight: 700;
    text-transform: capitalize;
    margin-bottom: 0.4rem;
  }

  .form-group input {
    color: white;
    font-size: 1.2rem;
    width: 100%;
    padding: 0.2rem 0.2rem;
    border-radius: 3px;
    border: 1px solid white;
    background-color: rgba(0, 0, 0, 0.1);
  }

  .login-form button[type='submit']  {
    cursor:pointer;
    color: black;
    width: 5em;
    height: 2em;
    margin: 1em;
    padding: 1em
    font-size: 6em;
    border: none;
    border-radius: 1em;
    &:hover {
      background-color: #6AB1DE;
      color: white;
    }
  }
  .login-form button[type='button'] {
    cursor:pointer;
    color: black;
    width: 5em;
    height: 2em;
    margin: 1em;
    padding: 1em
    font-size: 6em;
    border: none;
    border-radius: 1em;
   &:hover {
      background-color: #6AB1DE;
      color: white;
    }
  }
`
export default StyledLogin
