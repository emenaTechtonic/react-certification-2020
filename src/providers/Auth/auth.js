import React, { useState, useEffect, createContext, useCallback } from 'react'
import { AUTH_STORAGE_KEY, AUTH_STORAGE_USER } from '../../utils/constants'
import { storage } from '../../utils/storage'

export const AuthContext = createContext(null)

export const AuthProvider = ({ children }) => {
  const [authenticated, setAuthenticated] = useState(false)

  useEffect(() => {
    const lastAuthState = storage.get(AUTH_STORAGE_KEY)
    const isAuthenticated = Boolean(lastAuthState)

    setAuthenticated(isAuthenticated)
  }, [])

  const login = useCallback((user) => {
    setAuthenticated(true)
    storage.set(AUTH_STORAGE_KEY, true)
    storage.set(AUTH_STORAGE_USER, user)
  }, [])

  const logout = useCallback((user) => {
    setAuthenticated(false)
    storage.set(AUTH_STORAGE_KEY, false)
    storage.set(AUTH_STORAGE_USER, user)
  }, [])

  return (
    <AuthContext.Provider value={{ login, logout, authenticated }}>
      {children}
    </AuthContext.Provider>
  )
}

// export { useAuth }
export default AuthProvider
