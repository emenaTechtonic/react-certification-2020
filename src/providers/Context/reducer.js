export const ContextReducer = (state, action) => {
  switch (action.type) {
    case 'TOGGLE_THEME':
      return {
        ...state,
        theme: !state.theme,
      }
    case 'OPEN_MENU':
      return {
        ...state,
        menu: action.payload,
      }
    case 'GET_DATA':
      return {
        ...state,
        data: action.payload,
      }
    case 'SET_VIEW':
      return {
        ...state,
        view: action.payload,
      }
    case 'SET_TARGET':
      return {
        ...state,
        target: action.payload,
      }
    case 'SET_USER':
      return {
        ...state,
        user: action.payload,
      }
    case 'SET_FAVORITES':
      return {
        ...state,
        favorites: action.payload,
      }

    default:
      return state
  }
}
