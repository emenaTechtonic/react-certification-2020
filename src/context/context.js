import React, { useEffect, createContext, useReducer, memo } from 'react'
import { useFetch } from '../utils/hooks/useFetch'
import { ContextReducer } from './reducer'

export const Context = createContext(null)

export const initialState = {
  target: 'Wizeline',
  theme: false,
  menu: false,
  view: 'home',
  data: [],
}
export const ContextProvider = ({ children }) => {
  const [state, dispatch] = useReducer(ContextReducer, initialState)
  const { data, isLoading } = useFetch(state.target)

  useEffect(() => {
    if (data) dispatch({ type: 'GET_DATA', payload: data })
  }, [data])

  return isLoading ? (
    <div> Loading...</div>
  ) : (
    <Context.Provider value={{ state, dispatch }}> {children} </Context.Provider>
  )
}
export default memo(ContextProvider)
