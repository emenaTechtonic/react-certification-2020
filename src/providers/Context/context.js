import React, { useEffect, createContext, useCallback, useReducer, memo } from 'react'
import { useFetch } from '../../utils/hooks/useFetch'
import { storage } from '../../utils/storage'
import { isElementIn, removeElement } from '../../utils/fns'
import {
  AUTH_STORAGE_USER,
  AUTH_STORAGE_KEY,
  AUTH_STORAGE_FAVORITES,
} from '../../utils/constants'
import { ContextReducer } from './reducer'

export const Context = createContext(null)

export const initialState = {
  target: 'Wizeline',
  theme: false,
  menu: false,
  view: 'home',
  data: [],
  user: { id: 0, name: '', avatarUrl: '' },
  favorites: [],
}
export const ContextProvider = ({ children }) => {
  const [state, dispatch] = useReducer(ContextReducer, initialState)
  const { data } = useFetch(state.target)

  useEffect(() => {
    if (data) dispatch({ type: 'GET_DATA', payload: data })
  }, [data])
  useEffect(() => {
    const lastAuthState = storage.get(AUTH_STORAGE_KEY)
    const isAuthenticated = Boolean(lastAuthState)

    if (isAuthenticated) {
      const lastUser = storage.get(AUTH_STORAGE_USER)
      const userAuthenticated = Object(lastUser)

      dispatch({ type: 'SET_USER', payload: userAuthenticated })

      const lastFavorites = storage.get(AUTH_STORAGE_FAVORITES)
      if (lastFavorites) {
        const favorites = lastFavorites
        dispatch({ type: 'SET_FAVORITES', payload: favorites })
      }
    }
  }, [])

  const saveFavoriteCard = useCallback((card) => {
    let favorites = []
    const lastFavorites = storage.get(AUTH_STORAGE_FAVORITES)
    if (lastFavorites) {
      favorites = lastFavorites
    }
    if (card) {
      const res = isElementIn(card.id, favorites)
      if (!res) {
        favorites.push(card)
        dispatch({ type: 'SET_FAVORITES', payload: favorites })
        storage.set(AUTH_STORAGE_FAVORITES, favorites)
      }
    }
  }, [])
  const removeFavoriteCard = useCallback((card) => {
    let favorites = []
    const lastFavorites = storage.get(AUTH_STORAGE_FAVORITES)
    if (lastFavorites) {
      favorites = lastFavorites
    }
    if (card) {
      const newFavorites = removeElement(card.id, favorites)
      dispatch({ type: 'SET_FAVORITES', payload: newFavorites })
      storage.set(AUTH_STORAGE_FAVORITES, newFavorites)
    }
  }, [])

  // return isLoading ? (
  //   <div> Loading...</div>
  // ) : (
  return (
    <Context.Provider value={{ state, dispatch, saveFavoriteCard, removeFavoriteCard }}>
      {children}
    </Context.Provider>
  )
}
export default memo(ContextProvider)
