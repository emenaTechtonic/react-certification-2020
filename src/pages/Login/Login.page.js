import React, { useState, useContext } from 'react'
import debounce from 'lodash/debounce'
import { useHistory } from 'react-router'
import loginApi from '../../utils/assets/login.api'
import { Context } from '../../providers/Context/context'
import { AuthContext } from '../../providers/Auth/auth'
import StyledLogin from './Login.styled'

const styles = {
  popover: {
    position: 'absolute',
    width: 400,
    height: 400,
  },
}

function LoginPage({ coords, closeForm }) {
  const closeLogin = debounce(closeForm)
  const { dispatch } = useContext(Context)
  const [username, setUserName] = useState()
  const [password, setPassword] = useState()
  const [status, setStatus] = useState('Welcome back!')
  const { login } = useContext(AuthContext)
  const history = useHistory()

  const authenticate = async (e) => {
    e.preventDefault()
    try {
      const user = await loginApi({
        username,
        password,
      })
      if (user) {
        login(user)
        dispatch({ type: 'SET_USER', payload: user })
        history.push('/')
      }
    } catch (err) {
      setStatus('Login failed')
    }
  }

  return (
    <StyledLogin style={{ ...styles.popover, ...coords }}>
      <div className="login-wrapper">
        <h1>{status}</h1>
        <form onSubmit={authenticate} className="login-form">
          <div className="form-group">
            <label htmlFor="username">
              <strong>username </strong>
              <input
                required
                type="text"
                id="username"
                onChange={(e) => setUserName(e.target.value)}
              />
            </label>
          </div>
          <div className="form-group">
            <label htmlFor="password">
              <strong>password </strong>
              <input
                required
                type="password"
                id="password"
                onChange={(e) => setPassword(e.target.value)}
              />
            </label>
          </div>
          <div>
            <button type="submit">login</button>
            <button type="button" onClick={closeLogin}>
              cancel
            </button>
          </div>
        </form>
      </div>
    </StyledLogin>
  )
}

export default LoginPage
