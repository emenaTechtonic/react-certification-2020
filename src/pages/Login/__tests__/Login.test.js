import React from 'react'
import renderer from 'react-test-renderer'
import { Context } from '../../../providers/Context/context'
import { AuthProvider } from '../../../providers/Auth/auth'
import Login from '../Login.page'

const _ = require('lodash')

const state = {
  target: 'Wizeline',
  theme: false,
  data: [],
  user: { id: 0, name: '', avatarUrl: '' },
  favorites: [],
}
const login = null
const logout = null

describe('Testing Login component', () => {
  test('renders Login without crashing', async () => {
    const func = jest.fn()
    const debouncedFunc = _.debounce(func, 1000)
    const rendered = renderer.create(
      <AuthProvider authenticated={false} login={login} logout={logout}>
        <Context.Provider value={{ state }}>
          <Login closeForm={debouncedFunc} />
        </Context.Provider>
      </AuthProvider>
    )

    expect(rendered).toMatchSnapshot()
  })
})
