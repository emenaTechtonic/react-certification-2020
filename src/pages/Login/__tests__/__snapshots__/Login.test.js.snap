// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Testing Login component renders Login without crashing 1`] = `
<div
  className="sc-bdvvaa gTCtOE"
  role="login"
  style={
    Object {
      "height": 400,
      "position": "absolute",
      "width": 400,
    }
  }
>
  <div
    className="login-wrapper"
  >
    <h1>
      Welcome back!
    </h1>
    <form
      className="login-form"
      onSubmit={[Function]}
    >
      <div
        className="form-group"
      >
        <label
          htmlFor="username"
        >
          <strong>
            username 
          </strong>
          <input
            id="username"
            onChange={[Function]}
            required={true}
            type="text"
          />
        </label>
      </div>
      <div
        className="form-group"
      >
        <label
          htmlFor="password"
        >
          <strong>
            password 
          </strong>
          <input
            id="password"
            onChange={[Function]}
            required={true}
            type="password"
          />
        </label>
      </div>
      <div>
        <button
          type="submit"
        >
          login
        </button>
        <button
          onClick={[Function]}
          type="button"
        >
          cancel
        </button>
      </div>
    </form>
  </div>
</div>
`;
